### PR TITLE
Service Konto backend

### PR DESCRIPTION
**Pull request by 2e2a**
Testing without Frontend:

- in `etc/development.ini`, set `adhocracy.service_konto.enabled = true`.
- Create a Service Konto-Account: [https://skbref.verwalt-berlin.de/skb/FVP/Login/Anmeldung.aspx](https://skbref.verwalt-berlin.de/skb/FVP/Login/Anmeldung.aspx)
- Login to Service-Konto and select mein.berlin Sevice, then "weiter". You will be redirected to mein.berlin.de with a token as a parameter.
- Extract the Token from the network tab of the browser. 
- Login with curl: `curl http://localhost:6541/login_service_konto -H 'Content-Type: application/json;charset=UTF-8' --data-binary '{"token":"TOKEN"}`
- Note that the token is valid only for 20 seconds

### CODE REVIEW

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    result = service_konto.service.GetUserData(token)
+    if not result.GetUserDataResult == SERVICE_KONTO_GET_USER_DATA_SUCCESS:
+        raise ValueError('Failed getting user data (return_code: {}).'
+                         .format(result.GetUserDataResult))
+    return result.strXMLUserData
+
+
+def _parse_user_data_xml(xml: str) -> dict:
+    root = etree.fromstring(xml)
+    hhgw = _get_xml_element(root, 'HHGW')
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: What's HHGW?

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    user_data = dict()
+    required_attributes = ['USERID', 'FIRSTNAME', 'LASTNAME', 'EMAIL']
+    for attribute in required_attributes:
+        user_data[attribute.lower()] = hhgw.get(attribute)
+    if not _is_user_data_valid(user_data):
+        raise ValueError('Invalid user data received.')
+    return user_data
+
+
+def _get_xml_element(root: etree.ElementTree, tag: str) -> etree.Element:
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: One quick annoying question: Is this is an efficient way to find an xml element in a tree?

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+              userid: int) -> IUser:
+    locator = registry.getMultiAdapter((context, request), IUserLocator)
+    user = locator.get_user_by_service_konto_userid(userid)
+    return user
+
+
+def _create_user(context: IResource, registry: Registry, request: Request,
+                 user_data: dict) -> IUser:
+    if _is_email_used(context, registry, request, user_data.get('email')):
+        raise ValueError('User with ServiceKonto email already exists.')
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: To get things right: this error is thrown when a user has attempted to log in to meinberlin with a Servicekonto account whose email address is already used in a3, correct? If so, could you make the error message more descriptive?

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def _generate_username(context: IResource, registry: Registry,
+                       request: Request, user_data: dict) -> str:
+    name = user_data.get('firstname') + ' ' + user_data.get('lastname')
+    unique_name = name
+    i = 1
+    while _is_username_used(context, registry, request, unique_name):
+        unique_name = name + ' {}'.format(i)
+        i += 1
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: funny.

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py (hunk 3 above)
Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: OK

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py (hunk 1 above)
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: Yes. But why is it called that?

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py (hunk 2 above)
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: Cool.

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py (hunk 1 above)
Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: I have no idea and dont care :)

File: src/adhocracy_core/adhocracy_core/rest/schemas.py [hunk 5]
```diff
@@ -1298,3 +1297,31 @@ class POSTPasswordResetRequestSchema(MappingSchema):
                     validator=validate_password_reset_path,
                     )
     password = Password(missing=required)
+
+class POSTLoginServiceKontoSchema(MappingSchema):
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: Did this mean that `POSTLoginServiceKontoSchema` extends `MappingSchema`?

File: src/adhocracy_core/adhocracy_core/rest/schemas.py [hunk 6]
```diff
@@ -1298,3 +1297,31 @@ class POSTPasswordResetRequestSchema(MappingSchema):
...
+    """Schema for login requests via service konto token."""
+
+    token = SingleLine(missing=required)
+
+    @deferred
+    def validator(node: SchemaNode, kw: dict) -> All:
+        request = kw['request']
+        context = kw['context']
+        registry = kw['registry']
+        return All(
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: what's `All`? (it's hard to find something starting with a capital.)

File: src/adhocracy_core/adhocracy_core/rest/schemas.py (hunk 5 above)
Comment (xi) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: yes

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        raise ValueError('User with ServiceKonto email already exists.')
+    users = find_service(context, 'principals', 'users')
+    name = _generate_username(context, registry, request, user_data)
+    appstruct = {
+        IUserBasic.__identifier__: {'name': name},
+        IUserExtended.__identifier__: {'email': user_data.get('email')},
+        IServiceKonto.__identifier__: {'userid': int(user_data.get('userid'))},
+    }
+    user = registry.content.create(IUser.__identifier__, users, appstruct,
+                                   registry=registry, send_event=False)
```
Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: Not directly. I means that no internal event is send, which would be handled by the internal subscribers.

File: src/adhocracy_core/adhocracy_core/rest/schemas.py (hunk 5 above)
Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: yes

File: src/adhocracy_core/adhocracy_core/rest/schemas.py (hunk 6 above)
Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: Its a "Composite validator" imported from colander.

File: src/adhocracy_core/adhocracy_core/authentication/service_konto.py (hunk 1 above)
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/authentication/service_konto.py]: :)

File: src/adhocracy_core/adhocracy_core/rest/schemas.py (hunk 6 above)
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: got it.

File: src/adhocracy_core/adhocracy_core/sheets/principal.py [hunk 8]
```diff
@@ -436,6 +436,26 @@ class ServiceKontoSchema(MappingSchema):
...
 
 
+class IServiceKontoSettings(ISheet):
+    """Marker interface for the ServiceKonto settings sheet."""
+
+
+class ServiceKontoSettingsSchema(MappingSchema):
+    """Data structure for public ServiceKonto user information."""
+
+    enabled = Boolean()
```
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/sheets/principal.py]: I find your choice of var names - `Settings` and `enabled` - interesting. Will there be any settings besides the current boolean that we'll deal with?

Comment (2e2a) [inline src/adhocracy_core/adhocracy_core/sheets/principal.py]: Theoretically, something like import settings.

Comment (der-john) [inline src/adhocracy_core/adhocracy_core/sheets/principal.py]: ok.

File: src/adhocracy_core/adhocracy_core/rest/schemas.py (hunk 5 above)
Comment (der-john) [inline src/adhocracy_core/adhocracy_core/rest/schemas.py]: thanks.

**Comment by 2e2a** [comment]
Reviewed and works on staging.

**Pull request merged**